### ISSUE TITLE
be a state machine

### ISSUE THREAD

**Issue by diasdavid** [libp2p/js-libp2p-switch#24]
I've been debating with myself for some hours now of how to solve a old problem in the most elegant way possible.

It resumes to this:

a) If a connection is dropped (for reasons), a socket will emit that error, if not catched, it will make the program crash. 

b) If a connection, that was upgraded to a stream multiplexer drops, we get the error on the connection + an error emitted in every single multiplexed conn that was created on top of that original connection. 

This is kind of annoying, because connections will drop at all the times we could do something hacky and capture all the errors at least once so that the process doesn't crash. However this doesn't solve the problem of handling a crash on a connection that holds several multiplexed connections between two peers, because the most dumb solution would be for the app or apps (bitswap, peer routing, record store, etc) realising that their connection broke and ask to dial again, making several points in the code to try to dial to the same peer again, and to make things worse,  when you support multiple transports, stream muxers, possible relays, etc, the level of madness scales to infinity.

Not saying that with a very complex piece of code that tests to every single edge case, this wouldn't be possible, but if we could find a elegant solution to describe the conn broke problem, it would be 🌟

**Comment by noffle** [comment]
I don't know enough about the implementations to say if this is the case or not, but one way of looking at upgraded connections is that the connection object represents a connection not *to its underlying destination* but rather *to the multiplexer*. Under this model, a transport-level disconnect becomes a decision made by the multiplexer: either report the error to its sub-streams (but a redial operation would just be an inexpensive signal to the multiplexer asking it to re-establish its connection to the underlying endpoint), or even refuse to pass on the error to the sub-streams (enabling mosh-like persistent connections over unreliable or intermittent transports).

**Comment by pgte** [comment]
Assuming that messages are idempotent, the way I would design the thing would be:

You will have to construct a stream that, underneath, reconnects if it needs to, providing a duplex stream interface to the outside world.
On top of this stream you can then place the multiplexer.

Like this, the multiplexer would never see the underlying connection disrupted (unless a serious error happened, or a maximum number of reconnect attempts was reached, etc.)

The problem with this is that the underlying transport now doesn't guarantee the delivery of messages because you could be reconnecting all the time. Depending on your protocol, this could be hard to manage (do you do request-response, pub/sub, events?). You could have to retry messages, and that would break the abstraction.

Because of this, you would need a transport (and probably also a heartbeat) protocol on top of the reconnect-enabled stream, with buffering.

Socket <=> Reconnect-enabled stream <=> transport protocol <=> multiplexer

All of this would require a lot of buffering, and if you can't do back-pressure, this leads to memory exhaustion.

**Comment by diasdavid** [comment]
After some braintrust, we came to the conclusion that the most elegant (simple + efficient) to go is to create 'dial locks', where if two points of the code try to dial to peer1, then the second dial request will wait for the 1st one to be completed and if a stream muxer upgrade happens, the second dial will just open a muxed conn.

Thank you @pgte @noffle, @soarez and @ngginever for the insight :)

**Comment by diasdavid** [comment]
Reuse a lot of the logic from go-libp2p-swarm, namely:

- dial backoff - avoid super charge at one node, this saves a node from getting blasted when DHT queries are happenng
- coalescing dials
- dial locks

https://github.com/libp2p/go-libp2p/blob/master/p2p/net/swarm/dial_sync.go

**Comment by diasdavid** [comment]
Related to https://github.com/libp2p/js-libp2p/issues/98

**Comment by jacobheun** [comment]
I was thinking as a part of this it would be good to have the underlying connections also be state machines to more easily handle transitions. I wrote down some of my initial thoughts as well as what a possible fsm might look like:

## Goal
A single switch instance should be able to handle n+1 connections to n+1 peers.
Each of those connections should be able to gracefully handle upgrades from basic connections,
to private connections, to encrypted connections, to muxed connections. The connections
should also gracefully handle errors, so that they can revert to a former successful state.

## Design

### Connection State Machines (CSMs)
The Switch itself will be a state machine, exposing itself to dependents. Internally it should
also manage each connection to a peer as a state machine. Each CSM will maintain knowledge of its 
state, as well as house the underlying [interface-connection](https://github.com/libp2p/interface-connection). 
Individual CSMs will be able to more easily handle transitions, both successful and errored.
Understanding the state of the connection (dialing, raw connection, encrypted, muxed, etc), will provide
more opportunities for the Switch to make informed decisions about the health of the connection.
Things like dial backoffs, dial locks, and coalescing dials, should be much easier to observe and manage.

#### CSM States
* Create a base connection
* Dial the base connection over available transports
  * If transports error, fall back to circuit
  * If circuit errors, Error
* If Private Network Enabled, Encrypt for Private
  * If error, Abort
* Encrypt the connection by performing secio handshake
  * If handshake fails, abort
* Upgrade to muxed connection
  * If error and no protocol, keep encrypted connection and stop
  * If error, use encrypted connection and continue
  * If no error, use muxed connection and continue
* Perform Protocol Handshake
  * If error, Error. Connection is retained. (Account for repeat failures)
* Full connection returned

**Stop** - no connection is returned via callback, but the current state is retained for that peer

**Abort** - return an error via callback, dont save the state for the peer (do we want to retain state to temporarily blacklist a peer?)

**Error** - return an error via callback, save the state for retries (the intent is to be less serious than an abort. is the distinction needed? Less flagrant errors could be handled more gracefully with dial backoffs, where as aborts could result in some kind of blacklisting.)

```js
const CSM = fsm('DIAL', {
  DIAL: { 
    abort: 'ABORT',
    // emit events for different transport dials?
    encrypt: 'ENCRYPT',
    error: 'ERROR'
  },
  PRIVATE: { // for private networks
    encrypt: 'ENCRYPT',
    abort: 'ABORT',
    error: 'ERROR'
  },
  ENCRYPT: {
    upgrade: 'UPGRADE',
    error: 'ERROR'
  },
  UPGRADE: {
    stop: 'STOP',
    handshake: 'PROTOCOL_HANDSHAKE'
  },
  PROTOCOL_HANDSHAKE: {
    connect: 'CONNECT'
    error: 'ERROR'
  },
  STOP: {
    dial: 'UPGRADE'
  },
  CONNECT: { }, // The full connection is established
  ABORT: { }, // A severe event occurred
  ERROR: { } // An error occurred, but future dials may be allowed
})
```

### Things to Consider

#### Deteriorating connections
Connections should gracefully handle backoff to previous states. For example, a failure to upgrade
to a muxed connection may be due to an intermittent failure. We want to be able to attempt to upgrade
the connection again, but we don't want to try indefinitely. If a maximum number of attempts are
tried, the underlying connection should be terminated entirely. Subsequent, fresh, dial attempts 
could re-establish that connection later.

#### Connection Upgrades
Connection upgrades will happen frequently between peers. There are various states the peers could be in that we need to handle. The Switch should be able to handle upgrades from the various states without needing to retry logic. For example, if we dial a connection and are able to establish a muxed connection, but we don't yet have a protocol set for that peer we want to maintain that connection. Once a protocol is established, we should be able to immediately perform a protocol handshake without going through the  process of checking each state of the connection, we should know the state.

**Comment by pgte** [comment]
@jacobheun I lover this approach. I have 2 questions:

* What do you think about modelling the "DISCONNECTED" state? That way you could reuse the state object.
* What do you think about modelling transient states, like "dialing" and ", so that you can easily handle double dials, etc? More generally, what do you think about modelling states and not actions? "DIALING", "ENCRYPTING", "ENCRYPTED", "UPGRADING", "UPGRADED", etc?

**Comment by jacobheun** [comment]
@pgte,  Yes to both.  I've created a more verbose version using proper states instead of actions. There's a lot going on, but I think it really helps with the clarity of states and provides more opportunity to perform actions and/or observe state. I also replaced `STOP` with `ENCRYPTED` as that is the appropriate state a connection should be in, if it fails to mux. Is this in line with the approach you were thinking?

```js
const CSM = fsm('DISCONNECTED', {
  DISCONNECTED: { // No active connections exist for the peer
    dial: 'DIALING'
  },
  DIALING: { 
    abort: 'ABORTED',
    // emit events for different transport dials?
    done: 'DIALED'
    error: 'ERRORED'
  },
  DIALED: {
    encrypt: 'ENCRYPTING',
    privatize: 'PRIVATIZING'
  },
  PRIVATIZING: { // for private networks
    done: 'PRIVATIZED',
    abort: 'ABORTED',
    error: 'ERRORED'
  },
  PRIVATIZED: {
    encrypt: 'ENCRYPTING'
  },
  ENCRYPTING: {
    done: 'ENCRYPTED',
    error: 'ERROR'
  },
  ENCRYPTED: {
    upgrade: 'UPGRADING'
  },
  UPGRADING: {
    stop: 'ENCRYPTED', // If upgrading is not ready, hold the encrypted connection
    done: 'UPGRADED'
  },
  UPGRADED: {
    handshake: 'HANDSHAKING_PROTOCOL'
  },
  HANDSHAKING_PROTOCOL: {
    done: 'CONNECTED'
    error: 'ERROR'
  },
  CONNECTED: { // The full connection is established
    disconnect: 'DISCONNECTING'
  },
  DISCONNECTING: { 
    done: 'DISCONNECTED'
  },
  ABORT: { }, // A severe event occurred
  ERROR: { // An error occurred, but future dials may be allowed
    disconnect: 'DISCONNECTING' // There could be multiple options here, but this is a likely action
  } 
})
```

**Comment by pgte** [comment]
@jacobheun Yes, the internal states are much clearer to me now, thank you!

**Comment by jacobheun** [comment]
Released with 0.41.0
